### PR TITLE
WIP: Fix getindex of cached arrays not following MemoryLayout

### PR DESCRIPTION
**Pull request by TSGut**
@dlfivefifty As discussed in [this PR](https://github.com/JuliaLinearAlgebra/InfiniteLinearAlgebra.jl/pull/125#discussion_r1173841087) I will try retaining bandedness of cached arrays. The workaround I have been using defines getindex to first resize, then getindex afterwards.

This doesn't address the issue I found in https://github.com/JuliaArrays/LazyArrays.jl/issues/247, which seems to be triggered by `display()`. 

Since BandedMatrices is not part of the dependencies of this package and we likely don't want to add it, it's a bit weird to write tests for this but I think `UpperTriangular` and `SparseArray` serve as a reasonable stand-in.

### CODE REVIEW

**Comment by TSGut** [comment]
Still have to test a lot of edge cases but this works now

```julia
julia> using BandedMatrices

julia> A = BandedMatrix(0 => zeros(10), 1 => ones(10))
11×11 BandedMatrix{Float64} with bandwidths (0, 1):
 0.0  1.0   ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅ 
  ⋅   0.0  1.0   ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅ 
  ⋅    ⋅   0.0  1.0   ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅ 
  ⋅    ⋅    ⋅   0.0  1.0   ⋅    ⋅    ⋅    ⋅    ⋅    ⋅ 
  ⋅    ⋅    ⋅    ⋅   0.0  1.0   ⋅    ⋅    ⋅    ⋅    ⋅ 
  ⋅    ⋅    ⋅    ⋅    ⋅   0.0  1.0   ⋅    ⋅    ⋅    ⋅ 
  ⋅    ⋅    ⋅    ⋅    ⋅    ⋅   0.0  1.0   ⋅    ⋅    ⋅ 
  ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅   0.0  1.0   ⋅    ⋅ 
  ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅   0.0  1.0   ⋅ 
  ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅   0.0  1.0
  ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅   0.0

julia> cache(A);

julia> A[1:4,1:7]
4×7 BandedMatrix{Float64} with bandwidths (0, 1):
 0.0  1.0   ⋅    ⋅    ⋅    ⋅    ⋅ 
  ⋅   0.0  1.0   ⋅    ⋅    ⋅    ⋅ 
  ⋅    ⋅   0.0  1.0   ⋅    ⋅    ⋅ 
  ⋅    ⋅    ⋅   0.0  1.0   ⋅    ⋅ 

julia> A[1:10,1:10]
10×10 BandedMatrix{Float64} with bandwidths (0, 1):
 0.0  1.0   ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅ 
  ⋅   0.0  1.0   ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅ 
  ⋅    ⋅   0.0  1.0   ⋅    ⋅    ⋅    ⋅    ⋅    ⋅ 
  ⋅    ⋅    ⋅   0.0  1.0   ⋅    ⋅    ⋅    ⋅    ⋅ 
  ⋅    ⋅    ⋅    ⋅   0.0  1.0   ⋅    ⋅    ⋅    ⋅ 
  ⋅    ⋅    ⋅    ⋅    ⋅   0.0  1.0   ⋅    ⋅    ⋅ 
  ⋅    ⋅    ⋅    ⋅    ⋅    ⋅   0.0  1.0   ⋅    ⋅ 
  ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅   0.0  1.0   ⋅ 
  ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅   0.0  1.0
  ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅   0.0
```

File: src/cache.jl [hunk 1]
```diff
@@ -95,16 +95,98 @@ end
 
 getindex(A::AbstractCachedMatrix, I::Integer) = A[Base._to_subscript_indices(A, I)...]
 
-@inline getindex(A::AbstractCachedMatrix, kr::AbstractUnitRange, jr::AbstractUnitRange) = layout_getindex(A, kr, jr)
-@inline getindex(A::AbstractCachedMatrix, kr::AbstractVector, jr::AbstractVector) = layout_getindex(A, kr, jr)
-@inline getindex(A::AbstractCachedMatrix, k::Integer, jr::AbstractVector) = layout_getindex(A, k, jr)
+@inline function getindex(A::AbstractCachedMatrix, k::Integer, jr::AbstractUnitRange)
```
Comment (dlfivefifty) [inline src/cache.jl:98]: This definition won't work since its possible that `jr` is infinite. here's an option:
```suggestion
function _isfinite_getindex(A, kr, jr, maxkr::Int, maxjr::Int)
     resizedata!(A, maxkr, maxjr)
     getindex(A.data, kr, jr)
end
_isfinite_getindex(A, kr, jr, _, _) = layout_getindex(A, kr, jr)

@inline getindex(A::AbstractCachedMatrix, k::Integer, jr::AbstractUnitRange) = _isfinite_getindex(A::AbstractCachedMatrix, k, jr, maximum(k), maximum(jr))
# TODO other getindex
```

**Comment by TSGut** [comment]
Okay seems to work for the infinite case too now when combined with LazyBandedMatrices and BandedMatrices. Tests for this should probably go into one of those two packages rather than here.

```julia
julia> du = [[1., 2.]; Fill(1.,∞)];

julia> d = [[1.,2.]; Fill(3.,∞)];

julia> dl = Fill(1.,∞);

julia> A = _BandedMatrix(Hcat(Vcat(1.,du),d,dl)', ℵ₀, 1, 1)
ℵ₀×ℵ₀ BandedMatrix{Float64} with bandwidths (1, 1) with data vcat((vcat(Float64, vcat(2-element Vector{Float64}, ℵ₀-element Fill{Float64, 1, Tuple{InfiniteArrays.OneToInf{Int64}}} with indices OneToInf()) with indices OneToInf()) with indices OneToInf())' with indices Base.OneTo(1)×OneToInf(), (vcat(2-element Vector{Float64}, ℵ₀-element Fill{Float64, 1, Tuple{InfiniteArrays.OneToInf{Int64}}} with indices OneToInf()) with indices OneToInf())' with indices Base.OneTo(1)×OneToInf(), 1×ℵ₀ adjoint(::Fill{Float64, 1, Tuple{InfiniteArrays.OneToInf{Int64}}}) with eltype Float64 with indices Base.OneTo(1)×OneToInf()) with indices Base.OneTo(3)×OneToInf() with indices OneToInf()×OneToInf():
 1.0  1.0   ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅   …  
 1.0  2.0  2.0   ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅      
  ⋅   1.0  3.0  1.0   ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅      
  ⋅    ⋅   1.0  3.0  1.0   ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅      
  ⋅    ⋅    ⋅   1.0  3.0  1.0   ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅      
  ⋅    ⋅    ⋅    ⋅   1.0  3.0  1.0   ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅   …  
  ⋅    ⋅    ⋅    ⋅    ⋅   1.0  3.0  1.0   ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅      
  ⋅    ⋅    ⋅    ⋅    ⋅    ⋅   1.0  3.0  1.0   ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅      
 ⋮                        ⋮                        ⋮                        ⋮                   ⋱  

julia> C = cache(A)
ℵ₀×ℵ₀ CachedArray{Float64, 2, BandedMatrix{Float64, Matrix{Float64}, Base.OneTo{Int64}}, BandedMatrix{Float64, ApplyArray{Float64, 2, typeof(vcat), Tuple{Adjoint{Float64, ApplyArray{Float64, 1, typeof(vcat), Tuple{Float64, ApplyArray{Float64, 1, typeof(vcat), Tuple{Vector{Float64}, Fill{Float64, 1, Tuple{InfiniteArrays.OneToInf{Int64}}}}}}}}, Adjoint{Float64, ApplyArray{Float64, 1, typeof(vcat), Tuple{Vector{Float64}, Fill{Float64, 1, Tuple{InfiniteArrays.OneToInf{Int64}}}}}}, Adjoint{Float64, Fill{Float64, 1, Tuple{InfiniteArrays.OneToInf{Int64}}}}}}, InfiniteArrays.OneToInf{Int64}}} with indices OneToInf()×OneToInf():
 1.0  1.0   ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅   …  
 1.0  2.0  2.0   ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅      
  ⋅   1.0  3.0  1.0   ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅      
  ⋅    ⋅   1.0  3.0  1.0   ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅      
  ⋅    ⋅    ⋅   1.0  3.0  1.0   ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅      
  ⋅    ⋅    ⋅    ⋅   1.0  3.0  1.0   ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅   …  
  ⋅    ⋅    ⋅    ⋅    ⋅   1.0  3.0  1.0   ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅      
  ⋅    ⋅    ⋅    ⋅    ⋅    ⋅   1.0  3.0  1.0   ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅      
 ⋮                        ⋮                        ⋮                        ⋮                   ⋱  

julia> C[1:10,1:10]
10×10 BandedMatrix{Float64} with bandwidths (1, 1):
 1.0  1.0   ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅ 
 1.0  2.0  2.0   ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅ 
  ⋅   1.0  3.0  1.0   ⋅    ⋅    ⋅    ⋅    ⋅    ⋅ 
  ⋅    ⋅   1.0  3.0  1.0   ⋅    ⋅    ⋅    ⋅    ⋅ 
  ⋅    ⋅    ⋅   1.0  3.0  1.0   ⋅    ⋅    ⋅    ⋅ 
  ⋅    ⋅    ⋅    ⋅   1.0  3.0  1.0   ⋅    ⋅    ⋅ 
  ⋅    ⋅    ⋅    ⋅    ⋅   1.0  3.0  1.0   ⋅    ⋅ 
  ⋅    ⋅    ⋅    ⋅    ⋅    ⋅   1.0  3.0  1.0   ⋅ 
  ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅   1.0  3.0  1.0
  ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅   1.0  3.0

julia> C[5:∞,1:∞]
ℵ₀×ℵ₀ view(::CachedArray{Float64, 2, BandedMatrix{Float64, Matrix{Float64}, Base.OneTo{Int64}}, BandedMatrix{Float64, ApplyArray{Float64, 2, typeof(vcat), Tuple{Adjoint{Float64, ApplyArray{Float64, 1, typeof(vcat), Tuple{Float64, ApplyArray{Float64, 1, typeof(vcat), Tuple{Vector{Float64}, Fill{Float64, 1, Tuple{InfiniteArrays.OneToInf{Int64}}}}}}}}, Adjoint{Float64, ApplyArray{Float64, 1, typeof(vcat), Tuple{Vector{Float64}, Fill{Float64, 1, Tuple{InfiniteArrays.OneToInf{Int64}}}}}}, Adjoint{Float64, Fill{Float64, 1, Tuple{InfiniteArrays.OneToInf{Int64}}}}}}, InfiniteArrays.OneToInf{Int64}}}, 5:∞, 1:∞) with eltype Float64 with indices OneToInf()×OneToInf():
  ⋅    ⋅    ⋅   1.0  3.0  1.0   ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅   …  
  ⋅    ⋅    ⋅    ⋅   1.0  3.0  1.0   ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅      
  ⋅    ⋅    ⋅    ⋅    ⋅   1.0  3.0  1.0   ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅      
  ⋅    ⋅    ⋅    ⋅    ⋅    ⋅   1.0  3.0  1.0   ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅      
  ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅   1.0  3.0  1.0   ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅      
  ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅   1.0  3.0  1.0   ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅   …  
  ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅   1.0  3.0  1.0   ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅      
  ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅   1.0  3.0  1.0   ⋅    ⋅    ⋅    ⋅    ⋅    ⋅      
 ⋮                        ⋮                        ⋮                        ⋮                   ⋱  
```

**Comment by dlfivefifty** [comment]
We should add a check of downstream packages:

https://github.com/JuliaLinearAlgebra/BandedMatrices.jl/blob/master/.github/workflows/downstream.yml

**Review by dlfivefifty — CHANGES_REQUESTED** [review]

File: src/cache.jl [hunk 2]
```diff
@@ -458,7 +467,3 @@ function cacheddata(V::SubArray)
     data = cacheddata(P)
     view(data, intersect.(axes(data), parentindices(V))...)
 end
-
-# Triangular
```
[the remark this answers is not in the document]
Comment (TSGut) [reply src/cache.jl:462]: I am rewriting it atm

**Comment by TSGut** [comment]
Okay so functionality wise this does everything I wanted out of it now, including keeping the banded structure when calling LowerTriangular(QL.factors)  instead of filling it into a lower triangular thing.

Now let's see how this downstream thing works, have never used that before

**Comment by TSGut** [comment]
@dlfivefifty Okay so from what I can gather the downstream workflow thing allows us to tell LazyArrays to run the tests in a package that depends on it, in this case perhaps LazyBandedMatices would be more appropriate? Or perhaps BandedMatrices? But to clarify, to make this make sense I would still have to add tests for the correct behavior inside of those packages as well, right?

File: src/cache.jl [hunk 3]
```diff
@@ -95,16 +95,67 @@ end
...
+end
+@inline function _isfinite_getindex(A::LowerTriangular{T, <:AbstractCachedMatrix{T}}, kr, jr, maxkr::Int, maxjr::Int) where T
+    nmax = max(maxkr,maxjr)
+    resizedata!(A.data, nmax, nmax)
+    getindex(LowerTriangular(A.data.data[1:nmax,1:nmax]), kr, jr)
+end
+@inline _isfinite_getindex(A::LowerTriangular{T, <:AbstractCachedMatrix{T}}, kr, jr, _, _) where T = layout_getindex(A, kr, jr)
+@inline _isfinite_getindex(A::UpperTriangular{T, <:AbstractCachedMatrix{T}}, kr, jr, _, _) where T = layout_getindex(A, kr, jr)
+
+@inline getindex(A::LowerTriangular{T, <:AbstractCachedMatrix{T}}, k::Integer, jr::AbstractUnitRange) where T = _isfinite_getindex(A, k, jr, k, maximum(jr))
```
Comment (dlfivefifty) [inline src/cache.jl:132]:
```suggestion
@inline getindex(A::AbstractTriangular{T, <:AbstractCachedMatrix{T}}, k::Integer, jr::AbstractUnitRange) where T = _isfinite_getindex(A, k, jr, k, maximum(jr))
```
(and below)

File: src/cache.jl [hunk 4]
```diff
@@ -95,16 +95,67 @@ end
...
+@inline getindex(A::AbstractCachedMatrix, k::Integer, ::Colon) = _isfinite_getindex(A, k, :, k, size(A,2))
+@inline getindex(A::AbstractCachedMatrix, kr::AbstractVector, ::Colon) = _isfinite_getindex(A, kr, :, maximum(kr), size(A,2))
+@inline getindex(A::AbstractCachedMatrix, kr::AbstractUnitRange, ::Colon) = _isfinite_getindex(A, kr, :, maximum(kr), size(A,2))
+@inline getindex(A::AbstractCachedMatrix, ::Colon, j::Integer) = _isfinite_getindex(A, :, j, size(A,1), j)
+@inline getindex(A::AbstractCachedMatrix, ::Colon, jr::AbstractVector) = _isfinite_getindex(A, :, jr, size(A,1), maximum(jr))
+@inline getindex(A::AbstractCachedMatrix, ::Colon, jr::AbstractUnitRange) = _isfinite_getindex(A, :, jr, size(A,1), maximum(jr))
+@inline getindex(A::AbstractCachedMatrix, ::Colon, ::Colon) = _isfinite_getindex(A, :, :, size(A,1), size(A,2))
+
+# Triangular case
+@inline function _isfinite_getindex(A::UpperTriangular{T, <:AbstractCachedMatrix{T}}, kr, jr, maxkr::Int, maxjr::Int) where T
```
Comment (dlfivefifty) [inline src/cache.jl:119]: What about `UnitUpperTriangular` and `UnitLowerTriangular`?

Just do a for loop:
```julia
for Tri in (:UnitUpperTriangular, :UpperTriangular, :UnitLowerTriangular, :LowerTriangular)
…
end
```

**Review by dlfivefifty — CHANGES_REQUESTED** [review]
Also increase the code coverage

**Comment by TSGut** [comment]
Okay I think this should do it for coverage (why do workflows on this repo need manual approval?)

**Review by dlfivefifty — APPROVED** [review]

**Comment by TSGut** [comment]
I don't actually have the rights to merge in this repo but assuming the tests and coverage pass on this version, I am happy for it to be merged. @dlfivefifty

**Comment by TSGut** [comment]
@dlfivefifty Can you merge this?
